### PR TITLE
Fix for the inaccuracies in debt reporting for havven holders.

### PR DESCRIPTION
**Pull request by intmainvoid**
Also added a bunch of tests.

### CODE REVIEW

File: test/Havven.js [hunk 1]
```diff
@@ -1164,18 +1203,41 @@ contract('Havven', async function(accounts) {
...
 
+		const balanceOfAccount1 = await nUSDContract.balanceOf(account1);
+
 		// Then try to burn them all. Only 10 nomins (and fees) should be gone.
-		await havven.burnNomins(nUSD, await nUSDContract.balanceOf(account1), { from: account1 });
-		assert.bnEqual(await nUSDContract.balanceOf(account1), amountReceived);
+		await havven.burnNomins(nUSD, balanceOfAccount1, { from: account1 });
+		const balanceOfAccount1AfterBurn = await nUSDContract.balanceOf(account1);
+
+		// console.log('##### txn', txn);
```
Comment (intmainvoid) [inline test/Havven.js]: Ordinarily, I wouldn't check this kind of code in, but it is a WIP still.

File: test/Havven.js [hunk 2]
```diff
@@ -796,12 +804,43 @@ contract('Havven', async function(accounts) {
...
+		);
+
 		// Give some HAV to account1
 		await havven.transfer(account1, toUnit('1000'), { from: owner });
 
 		// account1 should be able to issue
-		await havven.issueNomins(nUSD, web3.utils.toBN('1'), { from: account1 });
+		await havven.issueNomins(nUSD, web3.utils.toBN('4'), { from: account1 });
+
+		// const txn = await havven.issueNomins(nUSD, web3.utils.toBN('1'), { from: account1 });
```
Comment (intmainvoid) [inline test/Havven.js]: Ordinarily, I wouldn't check this kind of rubbish in, but it is a WIP still and this code is still handy to figure out what is going on.

File: contracts/Havven.sol [hunk 3]
```diff
@@ -6,6 +6,7 @@ FILE INFORMATION
 file:       Havven.sol
 version:    2.0
 author:     Kevin Brown
+            Gavin Conway
```
Comment (intmainvoid) [inline contracts/Havven.sol]: Thought I've now earned my stripes ;)

File: contracts/Havven.sol [hunk 4]
```diff
@@ -672,20 +673,27 @@ contract Havven is ExternStateToken {
         // What will the new total be including the new value?
         uint newTotalDebtIssued = safeAdd(hdrValue, totalDebtIssued);
 
-        // What is their percentage of the total debt?
-        uint debtPercentage = safeDiv_dec(hdrValue, newTotalDebtIssued);
+        // What is their percentage (as a high precision int) of the total debt?
+        uint debtPercentage = safeDiv_dec_round_high_precision(
+            decToHighPrecisionDec(hdrValue),
```
Comment (thekevinbrown) [inline contracts/Havven.sol]: We should test if this is strictly necessary or if passing the decs in without converting first produces the same result mathematically.

File: contracts/Havven.sol [hunk 5]
```diff
@@ -672,20 +673,27 @@ contract Havven is ExternStateToken {
...
+        uint delta = safeSub(HIGH_PRECISION_UNIT, debtPercentage);
 
         // How much existing debt do they have?
         uint existingDebt = debtBalanceOf(messageSender, "HDR");
          
         // And what does their debt ownership look like including this previous stake?
         if (existingDebt > 0) {
-            debtPercentage = safeDiv_dec(safeAdd(hdrValue, existingDebt), newTotalDebtIssued);
+            debtPercentage = safeDiv_dec_round_high_precision(
+                decToHighPrecisionDec(safeAdd(hdrValue, existingDebt)),
```
Comment (thekevinbrown) [inline contracts/Havven.sol]: Same here

File: contracts/Havven.sol [hunk 6]
```diff
@@ -788,12 +799,15 @@ contract Havven is ExternStateToken {
 
         // What percentage of the total debt are they trying to remove?
         uint totalDebtIssued = totalIssuedNomins("HDR");
-        uint debtPercentage = safeDiv_dec(debtToRemove, totalDebtIssued);
+        uint debtPercentage = safeDiv_dec_round_high_precision(
+            decToHighPrecisionDec(debtToRemove),
```
Comment (thekevinbrown) [inline contracts/Havven.sol]: Same here

File: contracts/Havven.sol [hunk 7]
```diff
@@ -804,14 +818,19 @@ contract Havven is ExternStateToken {
             // What percentage of the debt will they be left with?
             uint newDebt = safeSub(existingDebt, debtToRemove);
             uint newTotalDebtIssued = safeSub(totalDebtIssued, debtToRemove);
-            uint newDebtPercentage = safeDiv_dec(newDebt, newTotalDebtIssued);
+            uint newDebtPercentage = safeDiv_dec_round_high_precision(
+                decToHighPrecisionDec(newDebt),
```
Comment (thekevinbrown) [inline contracts/Havven.sol]: Same here

File: contracts/Havven.sol [hunk 8]
```diff
@@ -804,14 +818,19 @@ contract Havven is ExternStateToken {
...
+            // Store the debt percentage and debt ledger as high precision integers
             issuanceData[messageSender].initialDebtOwnership = newDebtPercentage;
             issuanceData[messageSender].debtEntryIndex = debtLedger.length;
         }
 
-        // Update our cumulative ledger
-        debtLedger.push(safeMul_dec(debtLedger[debtLedger.length - 1], delta));
+        // Update our cumulative ledger. This is also a high precision integer.
+        uint newDebtLedgerEntry = safeMul_dec_round_high_precision(debtLedger[debtLedger.length - 1], delta);
+        debtLedger.push(newDebtLedgerEntry);
```
Comment (thekevinbrown) [inline contracts/Havven.sol]: Probably better to do 
```suggestion
        debtLedger.push(
            safeMul_dec_round_high_precision(debtLedger[debtLedger.length - 1], delta);
        );
```

to save on gas of creating the variable here.

File: contracts/Havven.sol [hunk 9]
```diff
@@ -880,7 +900,11 @@ contract Havven is ExternStateToken {
         uint totalSystemValue = totalIssuedNomins(currencyKey);
 
         // Their debt balance is their portion of the total system value.
-        return safeMul_dec(totalSystemValue, currentDebtOwnership);
+        uint highPrecisionBalance = safeMul_dec_round_high_precision(
+            decToHighPrecisionDec(totalSystemValue),
```
Comment (thekevinbrown) [inline contracts/Havven.sol]: Same here, do we actually need to do the conversion first?

File: contracts/SafeDecimalMath.sol [hunk 10]
```diff
@@ -137,7 +143,59 @@ contract SafeDecimalMath {
...
+    function safeMul_dec_round_private(uint x, uint y, uint unit)
+        private
+        pure
+        returns (uint)
+    {
+        /* Divide by UNIT to remove the extra factor introduced by the product. */
+        uint quotientTimesTen = safeMul(x, y) / (unit / 10);
+
+        if (quotientTimesTen % 10 >= 5) {
+            quotientTimesTen = quotientTimesTen + 10;
```
Comment (thekevinbrown) [inline contracts/SafeDecimalMath.sol]:
```suggestion
            quotientTimesTen += 10;
```

Maybe?

File: contracts/SafeDecimalMath.sol [hunk 11]
```diff
@@ -180,6 +238,46 @@ contract SafeDecimalMath {
         return safeDiv(safeMul(x, UNIT), y);
     }
 
+    function safeDiv_dec_round_private(uint x, uint y, uint unit)
```
Comment (thekevinbrown) [inline contracts/SafeDecimalMath.sol]: Should document this like the other functions.

File: contracts/SafeDecimalMath.sol [hunk 12]
```diff
@@ -36,12 +37,17 @@ pragma solidity 0.4.25;
...
+    /* Number of decimal places in the representations. */
     uint8 public constant decimals = 18;
+    uint8 public constant highPrecisionDecimals = 27;
 
     /* The number representing 1.0. */
     uint public constant UNIT = 10 ** uint(decimals);
 
+    /* The number representing 1.0 for higher fidelity numbers. */
+    uint public constant HIGH_PRECISION_UNIT = 10 ** uint(highPrecisionDecimals);
+    uint private constant UNIT_TO_HIGH_PRECISION_UNIT_CONVERTER = 10 ** uint(highPrecisionDecimals - decimals);
```
Comment (thekevinbrown) [inline contracts/SafeDecimalMath.sol]:
```suggestion
    uint private constant UNIT_TO_HIGH_PRECISION_CONVERSION_FACTOR = 10 ** uint(highPrecisionDecimals - decimals);
```

Maybe?

File: test/Havven.js [hunk 13]
```diff
@@ -796,12 +804,43 @@ contract('Havven', async function(accounts) {
 
 	// Issuance
 
-	it('should be possible to issue a small amount of nomins', async function() {
+	// TODO: Adding the exchange rates together for the HDR exchange value increases the chance that the amount
+	//  to be issued will be < 1 (ie. the smallest unit possible). This means that the user will in effect end up issuing
+	//  0 as we can't do any amounts less than 1 in Solidity.
+	//  Also, it seems that SafeDecimalMath truncates and doesn't round. So, 0.78 becomes 0 instead of 1. Is this a problem?
```
Comment (thekevinbrown) [inline test/Havven.js]: It rounds now!

File: test/Havven.js [hunk 14]
```diff
@@ -897,8 +936,8 @@ contract('Havven', async function(accounts) {
 		// But there's a small rounding error.
 		// This is ok, as when the last person exits the system, their debt percentage is always 100% so
 		// these rounding errors don't cause the system to be out of balance.
-		assert.bnEqual(await havven.debtBalanceOf(account1, nUSD), toUnit('10.00000000000000002'));
-		assert.bnEqual(await havven.debtBalanceOf(account2, nUSD), toUnit('19.99999999999999998'));
+		assertBNClose(await havven.debtBalanceOf(account1, nUSD), toUnit('10'), '3');
```
Comment (thekevinbrown) [inline test/Havven.js]: I'd suggest that we just set the default on assertBNClose to '5' now then omit the final parameter here.

File: test/Havven.js [hunk 15]
```diff
@@ -929,8 +968,8 @@ contract('Havven', async function(accounts) {
 		// But there's a small rounding error.
 		// This is ok, as when the last person exits the system, their debt percentage is always 100% so
 		// these rounding errors don't cause the system to be out of balance.
-		assert.bnEqual(await havven.debtBalanceOf(account1, nUSD), toUnit('20'));
-		assert.bnEqual(await havven.debtBalanceOf(account2, nUSD), toUnit('19.99999999999999992'));
+		assertBNClose(await havven.debtBalanceOf(account1, nUSD), toUnit('20'), '3');
+		assertBNClose(await havven.debtBalanceOf(account2, nUSD), toUnit('20'), '3');
```
Comment (thekevinbrown) [inline test/Havven.js]: Same here.

File: test/Havven.js [hunk 16]
```diff
@@ -1221,25 +1282,136 @@ contract('Havven', async function(accounts) {
 			.sub(burntNominsPt2);
 
 		// TODO: The variance we are getting here seems suspect. Let's investigate
-		assertBNClose(debt, expectedDebt, '10000');
+		assertBNClose(debt, expectedDebt, '5');
```
Comment (thekevinbrown) [inline test/Havven.js]: And we can then remove the 5 here too.

File: test/Havven.js [hunk 17]
```diff
@@ -1760,9 +1913,13 @@ contract('Havven', async function(accounts) {
...
 		await havven.burnNomins(nUSD, debtBalance3, { from: account3 });
 
-		assert.equal(debtBalance1, web3.utils.toBN('0'));
-		assert.equal(debtBalance2, web3.utils.toBN('0'));
-		assert.equal(debtBalance3, web3.utils.toBN('0'));
+		const debtBalance1After = await havven.debtBalanceOf(account1, nUSD);
+		const debtBalance2After = await havven.debtBalanceOf(account2, nUSD);
+		const debtBalance3After = await havven.debtBalanceOf(account3, nUSD);
+
+		assert.bnEqual(debtBalance1After, web3.utils.toBN('0'));
```
Comment (thekevinbrown) [inline test/Havven.js]: `assert.bnEqual` doesn't actually require BN instances, so you could write this like this:

```suggestion
		assert.bnEqual(debtBalance1After, '0');
```

File: test/Havven.js [hunk 18]
```diff
@@ -1760,9 +1913,13 @@ contract('Havven', async function(accounts) {
...
+		assert.bnEqual(debtBalance2After, web3.utils.toBN('0'));
```
Comment (thekevinbrown) [inline test/Havven.js]:
```suggestion
		assert.bnEqual(debtBalance2After, '0');
```

File: test/Havven.js [hunk 19]
```diff
@@ -1760,9 +1913,13 @@ contract('Havven', async function(accounts) {
...
+		assert.bnEqual(debtBalance3After, web3.utils.toBN('0'));
```
Comment (thekevinbrown) [inline test/Havven.js]:
```suggestion
		assert.bnEqual(debtBalance3After, '0');
```

File: test/Havven.js [hunk 20]
```diff
@@ -1780,9 +1937,11 @@ contract('Havven', async function(accounts) {
...
 		await havven.issueNomins(nUSD, issuedNomins3, { from: account3 });
 
-		const debtBalance = await havven.debtBalanceOf(account1, nUSD);
-		await havven.burnNomins(nUSD, debtBalance, { from: account1 });
-		assert.equal(debtBalance, web3.utils.toBN('0'));
+		const debtBalanceBefore = await havven.debtBalanceOf(account1, nUSD);
+		await havven.burnNomins(nUSD, debtBalanceBefore, { from: account1 });
+		const debtBalanceAfter = await havven.debtBalanceOf(account1, nUSD);
+
+		assert.bnEqual(debtBalanceAfter, web3.utils.toBN('0'));
```
Comment (thekevinbrown) [inline test/Havven.js]:
```suggestion
		assert.bnEqual(debtBalanceAfter, '0');
```

File: contracts/Havven.sol (hunk 6 above)
Comment (intmainvoid) [reply contracts/Havven.sol]: In this case it is important that the result of the calculation is in high precision. This is why we are "casting" the operands to high precision before doing the division.

File: contracts/Havven.sol (hunk 7 above)
Comment (intmainvoid) [reply contracts/Havven.sol]: In this case it is important that the result of the calculation is in high precision. This is why we are "casting" the operands to high precision before doing the division.

File: contracts/Havven.sol (hunk 9 above)
Comment (intmainvoid) [reply contracts/Havven.sol]: In the other cases, no. But in this case we do

**Pull request merged**